experimental/Schemes/Resolution_structure.jl

```diff
@@ -0,0 +1,738 @@
+# Desingularization morphism: birational map between covered schemes with smooth domain
+#####################################################################################################
+
+# Fehlt: NormalizationMorphism fuer Schemata -- muessten wir haben, sobald wir Lipman machen wollen
+#
+#@attributes mutable struct LipmanStyleSequence{
+#    DomainType<:AbsCoveredScheme,
+#    CodomainType<:AbsCoveredScheme
+#   } <: AbsDesingMor{
+#                                 DomainType,
+#                                 CodomainType,
+#                    }
+#  maps::Vector{:<union{BlowupMorphism,NormalizationMorphism}}        # count right to left:
+#                                                 # original scheme is codomain of map 1
+#  # boolean flags
+#  resolves_sing::Bool                            # domain not smooth yet?
+#  is_trivial::Bool                               # codomain already smooth?
+#
+#  # fields for caching, may be filled during computation
+#  ex_div::Vector{<:EffectiveCartierDivisor}      # list of exc. divisors arising from individual steps
+                                                  # in domain(maps[end])
+#
+#  # fields for caching to be filled a posteriori (on demand, only if partial_res==false)
+#  composed_map::AbsCoveredSchemeMorphism        
+#  exceptional_divisor::WeilDivisor          
+#
+#  function LipmanStyleSequence(maps::Vector{<:AbsCoveredSchemeMorphism})
+#    n = length(maps)
+#    for i in 1:n-1
+#      @assert domain(maps[i]) === codomain(maps[i+1]) "not a sequence of morphisms"
+#    end
+#    return new{typeof(domain(maps[end])),typeof(codomain(first(maps)))}(maps)
+#  end
+#end
+
+
+##################################################################################################
+# getters
+##################################################################################################
+maps(phi::AbsDesingMor) = copy(phi.maps)
+last_map(phi::AbsDesingMor) = phi.maps[end]
+exceptional_divisor_list(phi::BlowUpSequence) = phi.ex_div  ## derzeit Liste von Eff. Cartier Div.
+embeddings(phi::BlowUpSequence) = phi.embeddings
+
+## do not use!!! (for forwarding and certain emergenies)
+function underlying_morphism(phi::AbsDesingMor)
+  if !isdefined(phi, :composed_map)
+    len=length(maps(phi))
+    result=underlying_morphism(maps(phi)[1])
+    for i in 2:len
+      result = compose(underlying_morphism(maps(phi)[i]), result)
+    end
+    phi.composed_map = result
+  end
+  return phi.composed_map
+end
+
+##################################################################################################
+# setting values in DesingMors -- Watch out: only place with direct access to fields!!!
+##################################################################################################
+function add_map!(f::BlowUpSequence, phi::BlowupMorphism)
+  push!(f.maps, phi)
+  ex_div = [strict_transform(phi,E) for E in f.ex_div[1:end]]
+  push!(ex_div, Oscar.exceptional_divisor(phi))
```

Comment:
Is the `Oscar.` necessary? If yes, should `exceptional_divisor` be exported (resp. imported here)?
```suggestion
  push!(ex_div, exceptional_divisor(phi))
```

Reply:
done